Data Types tasks

dakaskos opened the pull request:
1. Creates sort function
2. Creates pick function
3. Creates omit function

dakaskos committed:
1. Creates sort function
2. Creates pick function
3. Creates omit function

jsru commented:
Добавляю преподавателя (@dosandk) для код-ревью.

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,19 @@
  */
 export function sortStrings(arr, param = 'asc') {
+  const sortedArr = [...arr].sort(function (a, b) {
+    if (a[0].match(/[a-z]/i) && !b[0].match(/[a-z]/i)) {
```
а что если кириллица? как это будет работать?

dakaskos replied:
Здесь представлено условие, что если символы из разных языков, то английская раскладка должна быть раньше. Если оба языка из одной раскладки, то делаем обычное сравнение

dakaskos replied:
Исправил на localeCompare with locale ['ru', 'en']

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,19 @@
+    return sortedArr;
+  }
+
+  return sortedArr.reverse();
```
Подумайте как исправить решение таким образом чтобы избавиться от `reverse` и сортировать данные сразу в верном порядке

dakaskos replied:
Умножить на -1 в самом sort. Исправил

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -5,5 +5,13 @@
 export const pick = (obj, ...fields) => {
+  const composedObj = {};
 
+  for (let key in fields) {
```
что вернет вот такой вызов:
`pick({}, 'constructor')`?
почему?

dakaskos replied:
Вернет магический метод constructor, так как лучше использовать Object.keys(obj), чтобы получить список ключей в виде списка. Исправил

dosandk commented on 02-javascript-data-types/3-omit/index.js:
```diff
@@ -5,5 +5,13 @@
 export const omit = (obj, ...fields) => {
+  const composedObj = {...obj};
 
+  for (let key in fields) {
```
массив стоит итерировать через `for...of`!
а для проверки наличия значения в массиве, метод `includes`

dakaskos replied:
Исправил

dosandk commented on 02-javascript-data-types/3-omit/index.js:
```diff
@@ -5,5 +5,13 @@
 
+  for (let key in fields) {
+    if (fields[key] in obj) {
+      delete composedObj[fields[key]];
```
использовать `delete` в связке с объектом не рекомендуется, ведь копирование объекта `{...obj}` делает `shallow copy` или другими словами неглубокую копию, поэтому удаление может привести к весьма неожиданным последствиям

dakaskos replied:
Исправил

dosandk reviewed (changes requested):
необходимо внести правки согласно комментариям

dakaskos committed:
1. Removes reverse from the sort
2. Use Object.keys instead of pure object
3. Update composed object by adding field instead of removing

jsru commented:
Добавляю преподавателя (@dosandk) для код-ревью.

dakaskos committed:
Use localeCompare with locale attribute ['ru', 'en']

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -7,9 +7,9 @@
-  for (let key in fields) {
-    if (fields[key] in obj) {
-      composedObj[fields[key]] = obj[fields[key]];
+  for (let field of fields) {
```
`const field`

dosandk reviewed (approved):
good 👍
